File: cogp-js/src/meta.ts
export const GEO_METADATA_KEY = 'geo';

export interface Level {
  row_group_end: number;
  resolution: number;
}

/** Fields every overview LoD declares, independent of its encoding. */
export interface OverviewLodBase {
  level_indices: number[];
  [field: string]: unknown;
}

/** A `quantized_geoarrow` LoD. */
export interface OverviewLod extends OverviewLodBase {
  geometry_type: 'LineString' | 'MultiLineString' | 'Polygon' | 'MultiPolygon';
  scale: [number, number];
  offset: [number, number];
}

export interface OverviewsMetadata {
  encoding: string;
  column: string;
  lods: Record<string, OverviewLodBase>;
}

export interface SupportedOverviewsMetadata extends OverviewsMetadata {
  encoding: 'quantized_geoarrow';
  lods: Record<string, OverviewLod>;
}

/** Identify encodings this reader can decode; parseLodMeta validates their fields. */
export function supportedOverviews(value: OverviewsMetadata | undefined): SupportedOverviewsMetadata | undefined {
  return value?.encoding === 'quantized_geoarrow' ? value as SupportedOverviewsMetadata : undefined;
}

/** The `geo.lod` object: feature-selection levels plus optional overviews. */
export interface LodMeta {
  levels: Level[];
  overviews?: OverviewsMetadata;
  [extra: string]: unknown;
}

export interface BboxCovering {
  xmin: string[];
  ymin: string[];
  xmax: string[];
  ymax: string[];
}

export interface Covering {
  bbox: BboxCovering;
}

export interface GeoColumn {
  encoding: string;
  geometry_types: string[];
  covering?: Covering;
  bbox?: number[];
  crs?: unknown;
  [extra: string]: unknown;
}

export interface GeoMeta {
  lod?: LodMeta;
  version: string;
  primary_column: string;
  columns: Record<string, GeoColumn>;
  [extra: string]: unknown;
}

export function parseLodMeta(json: string, numRowGroups?: number): LodMeta {
  const parsed = JSON.parse(json) as LodMeta;
  if (!parsed || typeof parsed !== 'object') throw new Error('geo.lod must be an object');
  if (!Array.isArray(parsed.levels) || parsed.levels.length === 0) {
    throw new Error('geo.lod: levels must be a non-empty array');
  }
  if (parsed.overviews !== undefined) {
    if (typeof parsed.overviews?.encoding !== 'string' || !parsed.overviews.encoding) {
      throw new Error('geo.lod: missing or invalid `overviews.encoding`');
    }
    const column = parsed.overviews.column;
    if (typeof column !== 'string' || !column) {
      throw new Error('geo.lod: missing or invalid `overviews.column`');
    }
    if (!parsed.overviews.lods || typeof parsed.overviews.lods !== 'object' || Array.isArray(parsed.overviews.lods)) {
      throw new Error('geo.lod: missing `overviews.lods`');
    }
    const lodEntries = Object.entries(parsed.overviews.lods);
    if (lodEntries.length === 0) {
      throw new Error('geo.lod: `overviews.lods` must be non-empty');
    }
    const assigned = new Set<number>();
    for (const [lod, metadata] of lodEntries) {
      if (!lod) throw new Error('geo.lod: invalid overview LoD name');
      const geometryType = metadata?.geometry_type;
      if (parsed.overviews.encoding === 'quantized_geoarrow'
        && (typeof geometryType !== 'string'
          || !['LineString', 'MultiLineString', 'Polygon', 'MultiPolygon'].includes(geometryType))) {
        throw new Error(`geo.lod: invalid geometry_type for overview ${lod}`);
      }
      if (!Array.isArray(metadata?.level_indices) || metadata.level_indices.length === 0) {
        throw new Error(`geo.lod: overview ${lod} level_indices must be a non-empty array`);
      }
      for (const index of metadata.level_indices) {
        if (!Number.isSafeInteger(index) || index < 0 || index >= parsed.levels.length) {
          throw new Error(`geo.lod: overview ${lod} level_indices index out of range`);
        }
        if (assigned.has(index)) throw new Error(`geo.lod: level ${index} is assigned more than once`);
        assigned.add(index);
      }
      if (supportedOverviews(parsed.overviews) && !validPair(metadata?.scale, true)) {
        throw new Error(`geo.lod: overviews.lods.${lod}.scale must contain two positive numbers`);
      }
      if (supportedOverviews(parsed.overviews) && !validPair(metadata?.offset, false)) {
        throw new Error(`geo.lod: overviews.lods.${lod}.offset must contain two finite numbers`);
      }
    }
    if (assigned.size !== parsed.levels.length) {
      throw new Error('geo.lod: every level must be assigned to an overview');
    }
  }
  let previousRowGroupEnd = -1;
  let previousResolution = Number.POSITIVE_INFINITY;
  for (const [index, level] of parsed.levels.entries()) {
    if (!level || typeof level !== 'object') throw new Error(`geo.lod: levels[${index}] must be an object`);
    if (!Number.isSafeInteger(level.row_group_end) || level.row_group_end < 0
      || level.row_group_end < previousRowGroupEnd) {
      throw new Error(`geo.lod: levels[${index}].row_group_end must be non-decreasing`);
    }
    if (!(Number.isFinite(level.resolution) && level.resolution > 0)) {
      throw new Error(`geo.lod: levels[${index}].resolution must be positive`);
    }
    if (level.resolution >= previousResolution) {
      throw new Error(`geo.lod: levels[${index}].resolution must strictly decrease`);
    }
    previousRowGroupEnd = level.row_group_end;
    previousResolution = level.resolution;
  }
  if (numRowGroups !== undefined && (!Number.isSafeInteger(numRowGroups) || numRowGroups <= 0 || previousRowGroupEnd !== numRowGroups - 1)) {
    throw new Error('geo.lod: final boundary must cover all row groups; empty files must omit the extension');
  }
  return parsed;
}

/** Metadata must be validated before resolving a level's rendering geometry. */
export function lodForLevel(metadata: LodMeta, index: number): string | undefined {
  return Object.entries(metadata.overviews?.lods ?? {})
    .find(([, lod]) => lod.level_indices.includes(index))?.[0];
}

function validPair(value: unknown, positive: boolean): value is [number, number] {
  return Array.isArray(value)
    && value.length === 2
    && value.every((item) => Number.isFinite(item) && (!positive || item > 0));
}

export function parseGeoMeta(json: string): GeoMeta {
  const parsed = JSON.parse(json) as GeoMeta;
  if (typeof parsed.primary_column !== 'string') {
    throw new Error('geo metadata: missing `primary_column`');
  }
  if (!parsed.columns || typeof parsed.columns !== 'object') {
    throw new Error('geo metadata: missing `columns`');
  }
  return parsed;
}

export type GeometryFamily = 'point' | 'line' | 'polygon';

export function geometryFamily(geometryTypes: readonly string[]): GeometryFamily | undefined {
  const family = (geometryType: string): GeometryFamily | undefined => {
    switch (geometryType.split(/\s+/, 1)[0]) {
      case 'Point':
      case 'MultiPoint': return 'point';
      case 'LineString':
      case 'MultiLineString': return 'line';
      case 'Polygon':
      case 'MultiPolygon': return 'polygon';
      default: return undefined;
    }
  };
  const first = geometryTypes[0] ? family(geometryTypes[0]) : undefined;
  return first && geometryTypes.every(type => family(type) === first) ? first : undefined;
}

export function extractGeoMeta(
  kv: ReadonlyArray<{ key: string; value?: string | null }> | null | undefined,
  numRowGroups: number,
): GeoMeta & { lod: LodMeta } {
  const geoJson = kv?.find(entry => entry.key === GEO_METADATA_KEY)?.value;
  if (!geoJson) throw new Error('not a GeoParquet file: missing `geo` key/value metadata');
  const geo = parseGeoMeta(geoJson);
  if (!geo.lod) throw new Error('missing geo.lod metadata');
  const lod = parseLodMeta(JSON.stringify(geo.lod), numRowGroups);
  const family = geometryFamily(geo.columns[geo.primary_column]?.geometry_types ?? []);
  if (supportedOverviews(lod.overviews) && family !== 'line' && family !== 'polygon') {
    throw new Error('overviews require a Line or Polygon family; points must not declare overviews');
  }
  if (lod.overviews) {
    if (lod.overviews.column === geo.primary_column) throw new Error('overview column must differ from primary geometry');
    for (const value of Object.values(supportedOverviews(lod.overviews)?.lods ?? {})) {
      if (geometryFamily([value.geometry_type]) !== family) {
        throw new Error('overview geometry_type must match primary geometry family');
      }
    }
  }
  return { ...geo, lod };
}

File: cogp-rs/src/meta.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

pub const GEO_METADATA_KEY: &str = "geo";
pub const GEOPARQUET_VERSION: &str = "1.1.0";
/// Default overview column name used by the converter; readers use `overviews.column`.
pub const OVERVIEWS_COLUMN: &str = "overviews";
pub const GEOARROW_ENCODING: &str = "quantized_geoarrow";

/// The `geo.lod` object: feature-selection levels plus optional overviews.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LodMeta {
    pub levels: Vec<Level>,
    #[serde(
        default,
        deserialize_with = "present_overviews",
        skip_serializing_if = "Option::is_none"
    )]
    pub overviews: Option<OverviewsMeta>,
}

impl LodMeta {
    /// Validate the layout and optional rendering contract before interpreting row ranges.
    /// Physical schema and per-row values are checked separately by the validator.
    pub fn validate(&self, num_row_groups: usize) -> anyhow::Result<()> {
        use anyhow::ensure;
        ensure!(!self.levels.is_empty(), "geo.lod.levels must be non-empty");
        ensure!(num_row_groups > 0, "file has zero row groups");
        let mut previous_end = -1;
        let mut previous_resolution = f64::INFINITY;
        for (index, level) in self.levels.iter().enumerate() {
            ensure!(
                level.row_group_end >= 0 && (level.row_group_end as u64) < num_row_groups as u64,
                "geo.lod.levels[{index}].row_group_end out of range"
            );
            ensure!(
                level.row_group_end >= previous_end,
                "geo.lod.levels[{index}].row_group_end must be non-decreasing"
            );
            ensure!(
                level.resolution.is_finite()
                    && level.resolution > 0.0
                    && level.resolution < previous_resolution,
                "geo.lod.levels[{index}].resolution must be positive, finite and strictly decreasing"
            );
            previous_end = level.row_group_end;
            previous_resolution = level.resolution;
        }
        ensure!(
            previous_end as u64 == num_row_groups as u64 - 1,
            "final geo.lod row_group_end must equal num_row_groups-1"
        );
        if let Some(overviews) = &self.overviews {
            ensure!(
                !overviews.encoding.is_empty(),
                "geo.lod.overviews.encoding must be non-empty"
            );
            ensure!(
                !overviews.column.is_empty(),
                "geo.lod.overviews.column must be non-empty"
            );
            ensure!(
                !overviews.lods.is_empty(),
                "geo.lod.overviews.lods must be non-empty"
            );
            let mut assigned = vec![false; self.levels.len()];
            for (lod, metadata) in &overviews.lods {
                ensure!(!lod.is_empty(), "invalid overview LoD name `{lod}`");
                if overviews.is_supported() {
                    let quantization = metadata
                        .quantization()
                        .map_err(|error| anyhow::anyhow!("overview `{lod}`: {error}"))?;
                    ensure!(
                        quantization.list_depth().is_some(),
                        "overview `{lod}` has invalid geometry_type"
                    );
                    ensure!(
                        quantization.scale.iter().all(|v| v.is_finite() && *v > 0.0),
                        "overview `{lod}` scale must be positive and finite"
                    );
                    ensure!(
                        quantization.offset.iter().all(|v| v.is_finite()),
                        "overview `{lod}` offset must be finite"
                    );
                }
                ensure!(
                    !metadata.level_indices.is_empty(),
                    "overview `{lod}` level_indices must be non-empty"
                );
                for &index in &metadata.level_indices {
                    ensure!(
                        index < assigned.len(),
                        "overview `{lod}` level_indices index {index} out of range"
                    );
                    ensure!(!assigned[index], "level {index} is assigned more than once");
                    assigned[index] = true;
                }
            }
            ensure!(
                assigned.iter().all(|value| *value),
                "every level must be assigned to an overview"
            );
        }
        Ok(())
    }

    /// Resolve rendering independently of the feature-selection level structure.
    pub fn lod_for_level(&self, index: usize) -> Option<&str> {
        self.overviews
            .as_ref()?
            .lods
            .iter()
            .find(|(_, metadata)| metadata.level_indices.contains(&index))
            .map(|(name, _)| name.as_str())
    }

    /// The shared LoD is present through the largest prefix that selects it.
    pub fn lod_row_group_end(&self, lod: &str) -> Option<i64> {
        self.overviews
            .as_ref()?
            .lods
            .get(lod)?
            .level_indices
            .iter()
            .filter_map(|&index| self.levels.get(index))
            .map(|level| level.row_group_end)
            .max()
    }
}

/// `overviews` may be omitted, but an explicit `null` is invalid metadata.
fn present_overviews<'de, D>(deserializer: D) -> Result<Option<OverviewsMeta>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    OverviewsMeta::deserialize(deserializer).map(Some)
}

/// JSON does not distinguish `2` from `2.0`; accept any integral safe integer.
fn integer<'de, D, T>(deserializer: D) -> Result<T, D::Error>
where
    D: serde::Deserializer<'de>,
    T: TryFrom<i64>,
{
    let number = serde_json::Number::deserialize(deserializer)?;
    let value = number
        .as_i64()
        .or_else(|| {
            number
                .as_f64()
                .filter(|v| v.fract() == 0.0 && v.abs() <= 9_007_199_254_740_991.0)
                .map(|v| v as i64)
        })
        .ok_or_else(|| serde::de::Error::custom(format!("expected an integer, got {number}")))?;
    T::try_from(value)
        .map_err(|_| serde::de::Error::custom(format!("integer {value} out of range")))
}

fn integers<'de, D>(deserializer: D) -> Result<Vec<usize>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    #[derive(Deserialize)]
    struct Index(#[serde(deserialize_with = "integer")] usize);
    Ok(Vec::<Index>::deserialize(deserializer)?
        .into_iter()
        .map(|Index(value)| value)
        .collect())
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Level {
    #[serde(deserialize_with = "integer")]
    pub row_group_end: i64,
    pub resolution: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OverviewsMeta {
    pub column: String,
    pub encoding: String,
    pub lods: BTreeMap<String, OverviewLod>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OverviewLod {
    #[serde(deserialize_with = "integers")]
    pub level_indices: Vec<usize>,
    // Only the encoding owns these fields. Unknown encodings remain opaque.
    #[serde(flatten)]
    pub properties: BTreeMap<String, serde_json::Value>,
}

impl OverviewLod {
    pub fn quantization(&self) -> anyhow::Result<QuantizationMeta> {
        Ok(serde_json::from_value(serde_json::to_value(
            &self.properties,
        )?)?)
    }
}

impl OverviewsMeta {
    /// Whether this implementation can decode the declared encoding.
    pub fn is_supported(&self) -> bool {
        self.encoding == GEOARROW_ENCODING
    }
}

/// Per-LoD fields defined by `quantized_geoarrow`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QuantizationMeta {
    pub geometry_type: String,
    pub scale: [f64; 2],
    pub offset: [f64; 2],
}

impl QuantizationMeta {
    pub fn list_depth(&self) -> Option<usize> {
        match self.geometry_type.as_str() {
            "LineString" => Some(1),
            "MultiLineString" | "Polygon" => Some(2),
            "MultiPolygon" => Some(3),
            _ => None,
        }
    }
}

/// GeoParquet file metadata. Only `primary_column` and `columns` are required
/// to read a file; the validator checks the remaining GeoParquet fields.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeoMeta {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lod: Option<LodMeta>,
    #[serde(default)]
    pub version: String,
    pub primary_column: String,
    pub columns: BTreeMap<String, GeoColumn>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeoColumn {
    #[serde(default)]
    pub encoding: String,
    #[serde(default)]
    pub geometry_types: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub covering: Option<Covering>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bbox: Option<Vec<f64>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub crs: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GeometryFamily {
    Point,
    Line,
    Polygon,
}

/// Rendering overviews use one topological geometry family. Singular and Multi
/// variants belong to the same family; dimensional suffixes such as `Z` and
/// `ZM` do not change it.
pub fn geometry_family(geometry_types: &[String]) -> Option<GeometryFamily> {
    fn one(geometry_type: &str) -> Option<GeometryFamily> {
        match geometry_type.split_ascii_whitespace().next()? {
            "Point" | "MultiPoint" => Some(GeometryFamily::Point),
            "LineString" | "MultiLineString" => Some(GeometryFamily::Line),
            "Polygon" | "MultiPolygon" => Some(GeometryFamily::Polygon),
            _ => None,
        }
    }

    let family = one(geometry_types.first()?)?;
    geometry_types
        .iter()
        .all(|geometry_type| one(geometry_type) == Some(family))
        .then_some(family)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Covering {
    pub bbox: BboxCovering,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BboxCovering {
    pub xmin: Vec<String>,
    pub ymin: Vec<String>,
    pub xmax: Vec<String>,
    pub ymax: Vec<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn unknown_encoding_preserves_opaque_metadata_and_validates_assignments() {
        let value = json!({"levels":[{"row_group_end":0,"resolution":1.0}],
        "overviews":{"column":"future","encoding":"future_v3","lods":{
            "all":{"level_indices":[0],"scale":{"future":"format"},"geometry_type":42}
        }}});
        let metadata: LodMeta = serde_json::from_value(value.clone()).unwrap();
        metadata.validate(1).unwrap();
        assert_eq!(serde_json::to_value(&metadata).unwrap(), value);
        let mut broken = metadata;
        broken
            .overviews
            .as_mut()
            .unwrap()
            .lods
            .get_mut("all")
            .unwrap()
            .level_indices = vec![1];
        assert!(broken.validate(1).is_err());
    }

    #[test]
    fn integral_json_numbers_are_integers_but_null_overviews_are_invalid() {
        let value = json!({"levels":[{"row_group_end":0.0,"resolution":1.0}],
        "overviews":{"column":"future","encoding":"future_v3","lods":{"all":{"level_indices":[0.0]}}}});
        let metadata: LodMeta = serde_json::from_value(value).unwrap();
        metadata.validate(1).unwrap();
        assert!(serde_json::from_value::<LodMeta>(
            json!({"levels":[{"row_group_end":0.5,"resolution":1.0}]})
        )
        .is_err());
        assert!(serde_json::from_value::<LodMeta>(
            json!({"levels":[{"row_group_end":0,"resolution":1.0}],"overviews":null})
        )
        .is_err());
    }

    #[test]
    fn geometry_family_accepts_singular_and_multi_but_rejects_mixed_dimensions() {
        assert_eq!(
            geometry_family(&["LineString".into(), "MultiLineString Z".into()]),
            Some(GeometryFamily::Line)
        );
        assert_eq!(
            geometry_family(&["LineString".into(), "Polygon".into()]),
            None
        );
        assert_eq!(geometry_family(&[]), None);
    }

    #[test]
    fn levels_are_independent_of_optional_overviews() {
        let metadata = LodMeta {
            levels: vec![Level {
                row_group_end: 0,
                resolution: 1000.0,
            }],
            overviews: None,
        };
        metadata.validate(1).unwrap();

        let serialized = serde_json::to_value(&metadata).unwrap();
        assert_eq!(
            serialized,
            json!({"levels": [{"row_group_end": 0, "resolution": 1000.0}]})
        );
    }

    #[test]
    fn overview_assignments_are_complete_unique_and_in_range() {
        let fixture: serde_json::Value =
            serde_json::from_str(include_str!("../tests/fixtures/metadata-overviews.json"))
                .unwrap();
        for indices in [json!([]), json!([4]), json!([0, 0]), json!([1])] {
            let mut value = fixture.clone();
            value["overviews"]["lods"]["l0"]["level_indices"] = indices;
            let metadata: LodMeta = serde_json::from_value(value).unwrap();
            assert!(metadata.validate(4).is_err());
        }
        for indices in [
            json!(null),
            json!(-1),
            json!([-1]),
            json!([0.5]),
            json!("0"),
        ] {
            let mut value = fixture.clone();
            value["overviews"]["lods"]["l0"]["level_indices"] = indices;
            assert!(serde_json::from_value::<LodMeta>(value).is_err());
        }
        let mut missing = fixture.clone();
        missing["overviews"]["lods"]
            .as_object_mut()
            .unwrap()
            .remove("l0");
        assert!(serde_json::from_value::<LodMeta>(missing)
            .unwrap()
            .validate(4)
            .is_err());
        let mut old = fixture;
        old["overviews"]["lods"]["l0"]
            .as_object_mut()
            .unwrap()
            .remove("level_indices");
        assert!(serde_json::from_value::<LodMeta>(old).is_err());
    }

    #[test]
    fn shared_overviews_use_explicit_indices_and_maximum_coverage() {
        let mut metadata: LodMeta =
            serde_json::from_str(include_str!("../tests/fixtures/metadata-overviews.json"))
                .unwrap();
        metadata
            .overviews
            .as_mut()
            .unwrap()
            .lods
            .get_mut("l1")
            .unwrap()
            .level_indices = vec![2, 1];
        metadata.validate(4).unwrap();
        assert_eq!(metadata.lod_for_level(1), Some("l1"));
        assert_eq!(metadata.lod_for_level(2), Some("l1"));
        assert_eq!(metadata.lod_row_group_end("l1"), Some(3));
        assert_eq!(metadata.lod_for_level(4), None);
        assert_eq!(metadata.lod_row_group_end("missing"), None);
        // Shared levels need not be adjacent, nor their indices sorted.
        let lods = &mut metadata.overviews.as_mut().unwrap().lods;
        lods.get_mut("l0").unwrap().level_indices = vec![2, 0];
        lods.get_mut("l1").unwrap().level_indices = vec![1];
        metadata.validate(4).unwrap();
        assert_eq!(metadata.lod_row_group_end("l0"), Some(3));
        let with_overviews = serde_json::to_value(&metadata).unwrap();
        metadata.overviews = None;
        metadata.validate(4).unwrap();
        assert_eq!(
            with_overviews["levels"],
            serde_json::to_value(&metadata).unwrap()["levels"]
        );
        assert_eq!(metadata.lod_for_level(0), None);
    }

    #[test]
    fn geo_meta_minimum_required_fields() {
        // Mimics the minimum a real GeoParquet writer would emit.
        let v = json!({
            "version": "1.1.0",
            "primary_column": "geometry",
            "columns": {
                "geometry": {
                    "encoding": "WKB",
                    "geometry_types": ["Polygon"],
                    "covering": {
                        "bbox": {
                            "xmin": ["bbox", "xmin"],
                            "ymin": ["bbox", "ymin"],
                            "xmax": ["bbox", "xmax"],
                            "ymax": ["bbox", "ymax"],
                        }
                    }
                }
            }
        });
        let parsed: GeoMeta = serde_json::from_value(v).unwrap();
        assert_eq!(parsed.primary_column, "geometry");
        let col = parsed.columns.get("geometry").unwrap();
        assert_eq!(col.encoding, "WKB");
        assert!(col.covering.is_some());
        assert!(col.bbox.is_none());
        assert!(col.crs.is_none());
    }

    #[test]
    fn geo_meta_optional_fields_skipped_on_serialize() {
        let col = GeoColumn {
            encoding: "WKB".into(),
            geometry_types: vec!["Polygon".into()],
            covering: None,
            bbox: None,
            crs: None,
        };
        let s = serde_json::to_string(&col).unwrap();
        // None values must be omitted, not serialized as null.
        assert!(!s.contains("covering"));
        assert!(!s.contains("bbox"));
        assert!(!s.contains("crs"));
    }
}
